File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolveParsedResultsPropertyName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node, context) => {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    const property = getParent(node)
    if (!property) return null
    if (!ts.isPropertySignature(property)) return null
    if (!ts.isIdentifier(property.name)) return null

    const propertyName = property.name.text

    const typeLiteral = getParent(property)
    if (!typeLiteral) return null
    if (!ts.isTypeLiteralNode(typeLiteral)) return null

    const conditionalType = getParent(typeLiteral)
    if (!conditionalType) return null
    if (!ts.isConditionalTypeNode(conditionalType)) return null

    const typeAlias = getParent(conditionalType)
    if (!typeAlias) return null
    if (!ts.isTypeAliasDeclaration(typeAlias)) return null

    const parentName = typeAlias.name.text
    if (parentName !== "ParsedResults") return null

    return `${karakum.capitalize(parentName)}${karakum.capitalize(propertyName)}`
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/nameResolvers/resolveParsedResultsPropertyName.kt
package node.karakum.nameResolvers

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import arrow.core.raise.nullable
import typescript.Node
import typescript.isConditionalTypeNode
import typescript.isIdentifier
import typescript.isPropertySignature
import typescript.isTypeAliasDeclaration
import typescript.isTypeLiteralNode

fun resolveParsedResultsPropertyName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val property = ensureNotNull(typeScriptService.getParent(node))
    ensure(isPropertySignature(property))

    val propertyNameNode = property.name
    ensure(isIdentifier(propertyNameNode))
    val propertyName = propertyNameNode.text

    val typeLiteral = ensureNotNull(typeScriptService.getParent(property))
    ensure(isTypeLiteralNode(typeLiteral))

    val conditionalType = ensureNotNull(typeScriptService.getParent(typeLiteral))
    ensure(isConditionalTypeNode(conditionalType))

    val typeAlias = ensureNotNull(typeScriptService.getParent(conditionalType))
    ensure(isTypeAliasDeclaration(typeAlias))
    val parentName = typeAlias.name.text
    ensure(parentName == "ParsedResults")

    "${parentName.replaceFirstChar { it.titlecase() }}${propertyName.replaceFirstChar { it.titlecase() }}"
}
